File: packages/python/sharing/starfish_sharing/public_link.py
"""Public-link API for plaintext (cap-only) sharing — Python mirror of
``public-link.ts``.

A public link is an ``audience`` cap-cert packed into a URL ``#fragment``.
Unlike a ``member`` cap it binds **no** single subject: every redeemer signs
requests with their **own** identity key (named via the ``X-Starfish-Pub``
header), so writes are attributable per user. An optional allow-list
(``allowed_identities``) narrows who may redeem; when omitted, any identity may.
No private key is ever embedded in the link.

``parse_public_link`` shape-checks the embedded cap but does NOT verify its
signature or expiry — that is the server's job at request time.
"""

from __future__ import annotations

import base64
import json
from dataclasses import dataclass
from typing import Any, Optional

from starfish_protocol.hash import stable_stringify
from starfish_protocol.request_signing import sign_request

from .cap_mint import AudienceMintOpts, ScopePreset, assert_audience_cap_shape, mint_audience_cap

# Current public-link payload version.
_PUBLIC_LINK_V = 1


@dataclass
class PublicLink:
    """Result of :func:`create_public_link`."""

    # base64url payload to place after the ``#`` in a share URL.
    fragment: str
    # The minted audience cap-cert (also embedded in ``fragment``).
    cap: dict[str, Any]


@dataclass
class ParsedPublicLink:
    """Result of :func:`parse_public_link`."""

    cap: dict[str, Any]


# ── base64url (URL-fragment-safe) ───────────────────────────────────────────────
# Standard base64 (``+``, ``/``, ``=``) is unsafe in a URL fragment; map to
# base64url and strip padding. The exact same mapping is mirrored byte-for-byte
# in the TS ``public-link.ts``.


def _b64url_encode(s: str) -> str:
    std = base64.urlsafe_b64encode(s.encode("utf-8")).decode("ascii")
    return std.rstrip("=")


def _b64url_decode(s: str) -> str:
    rem = len(s) % 4
    if rem == 2:
        s += "=="
    elif rem == 3:
        s += "="
    elif rem == 1:
        raise ValueError("malformed public link: bad base64url length")
    return base64.urlsafe_b64decode(s.encode("ascii")).decode("utf-8")


def _encode_cap_auth(cap: dict[str, Any]) -> str:
    """Encode a cap-cert for the ``Authorization: Cap <…>`` header (matches the resolver)."""
    return base64.b64encode(stable_stringify(cap).encode("utf-8")).decode("ascii")


def create_public_link(
    iss_ed_priv_hex: str,
    iss_ed_pub_hex: str,
    collection: str,
    scope: ScopePreset | dict[str, Any],
    *,
    allowed_identities: Optional[list[str]] = None,
    expires_at: Optional[int] = None,
    ttl_sec: Optional[int] = None,
    nbf: Optional[int] = None,
    nonce: Optional[bytes] = None,
) -> PublicLink:
    """Mint an audience cap and pack it into a shareable URL fragment.

    With ``allowed_identities`` the link works only for those identities;
    without it, any identity may redeem. Either way the link carries no private
    key.
    """
    cap = mint_audience_cap(
        iss_ed_priv_hex,
        iss_ed_pub_hex,
        collection,
        scope,
        AudienceMintOpts(
            audience=allowed_identities,
            expires_at=expires_at,
            ttl_sec=ttl_sec,
            nbf=nbf,
            nonce=nonce,
        ),
    )
    payload = {"v": _PUBLIC_LINK_V, "cap": cap}
    fragment = _b64url_encode(stable_stringify(payload))
    return PublicLink(fragment=fragment, cap=cap)


def parse_public_link(fragment: str) -> ParsedPublicLink:
    """Decode and shape-check a public-link fragment.

    Does NOT verify the cap's signature or expiry — the server does that at
    request time. Raises :class:`ValueError` on a malformed fragment, wrong
    payload version, or a non-audience / malformed cap.
    """
    try:
        payload = json.loads(_b64url_decode(fragment.strip()))
    except (ValueError, json.JSONDecodeError) as exc:
        raise ValueError("malformed public link: invalid fragment encoding") from exc
    if not isinstance(payload, dict):
        raise ValueError("malformed public link: payload is not an object")
    if payload.get("v") != _PUBLIC_LINK_V:
        raise ValueError(f"unsupported public link version: {payload.get('v')}")
    cap = payload.get("cap")
    if not isinstance(cap, dict) or cap.get("kind") != "audience":
        raise ValueError("malformed public link: cap is not an audience cap")
    # Structural validation only (no signature/expiry check — server's job).
    assert_audience_cap_shape(cap)
    return ParsedPublicLink(cap=cap)


def redeem_public_link(
    parsed: ParsedPublicLink,
    *,
    redeemer_ed_priv_hex: str,
    redeemer_ed_pub_hex: str,
    method: str,
    path_and_query: str,
    body: bytes | str = b"",
    host: Optional[str] = None,
    ts: Optional[int] = None,
    nonce: Optional[bytes] = None,
) -> dict[str, str]:
    """Build the header set a redeemer sends with each request, signing with the
    redeemer's own key and naming it via ``X-Starfish-Pub``.

    Transport-agnostic: the caller attaches these headers to its HTTP request.
    The body passed here MUST equal the bytes sent on the wire so the signature
    the server reconstructs matches.
    """
    body_bytes = body.encode("utf-8") if isinstance(body, str) else body
    signature = sign_request(
        method,
        path_and_query,
        body_bytes,
        redeemer_ed_priv_hex,
        host=host,
        ts=ts,
        nonce=nonce,
    )
    return {
        "Authorization": f"Cap {_encode_cap_auth(parsed.cap)}",
        "X-Starfish-Sig": signature.sig,
        "X-Starfish-Ts": str(signature.ts),
        "X-Starfish-Nonce": signature.nonce,
        "X-Starfish-Pub": redeemer_ed_pub_hex,
    }


__all__ = [
    "PublicLink",
    "ParsedPublicLink",
    "create_public_link",
    "parse_public_link",
    "redeem_public_link",
]

File: packages/ts/sharing/src/public-link.ts
/**
 * Public-link API for plaintext (cap-only) sharing.
 *
 * A public link is an `audience` cap-cert packed into a URL `#fragment`. Unlike
 * a `member` cap it binds **no** single subject: every redeemer signs requests
 * with their **own** identity key (named via the `X-Starfish-Pub` header), so
 * writes are attributable per user. An optional allow-list (`allowedIdentities`)
 * narrows who may redeem; when omitted, any identity may. No private key is ever
 * embedded in the link.
 *
 * Flow:
 *   owner:    `createPublicLink(...)`  → share `link#<fragment>`
 *   redeemer: `parsePublicLink(fragment)` → `redeemPublicLink(parsed, ...)`
 *             → send the returned headers with each request.
 *
 * `parsePublicLink` shape-checks the embedded cap but does NOT verify its
 * signature or expiry — that is the server's job at request time.
 */

import {
  getBase64,
  signRequest,
  stableStringify,
  type CapCert,
  type SignableMethod,
  type SignableRequest,
} from "@drakkar.software/starfish-protocol"
import {
  assertAudienceCapShape,
  mintAudienceCap,
  type ScopePreset,
} from "./cap-mint.js"

/** Current public-link payload version. */
const PUBLIC_LINK_V = 1

/** Options for {@link createPublicLink}. */
export interface CreatePublicLinkOpts {
  /** Issuer (collection owner) Ed25519 private key, hex. */
  issEdPrivHex: string
  /** Issuer Ed25519 public key, hex. */
  issEdPubHex: string
  /** Single collection the link grants access to. */
  collection: string
  /** Scope preset (e.g. `scopes.readOnly(col)` / `scopes.writer(col)`). */
  scope: ScopePreset
  /**
   * Optional allow-list of redeemer Ed25519 pubkeys (64-char lowercase hex).
   * Omit for "any identity may redeem"; provide to restrict to these identities.
   */
  allowedIdentities?: string[]
  /** Absolute expiry, unix seconds. Wins over `ttlSec`. Must be after `nbf`. */
  expiresAt?: number
  /** TTL in seconds from `nbf`. Defaults to 30 days when neither this nor `expiresAt` is set. */
  ttlSec?: number
  /** Not-before, unix seconds. Defaults to now. */
  nbf?: number
  /** Random nonce bytes (16 recommended). Defaults to fresh randomness. */
  nonce?: Uint8Array
}

/** Result of {@link createPublicLink}. */
export interface PublicLink {
  /** base64url payload to place after the `#` in a share URL. */
  fragment: string
  /** The minted audience cap-cert (also embedded in `fragment`). */
  cap: CapCert
}

/** Result of {@link parsePublicLink}. */
export interface ParsedPublicLink {
  /** The audience cap-cert carried by the link. */
  cap: CapCert
}

/** Options for {@link redeemPublicLink}. */
export interface RedeemPublicLinkOpts {
  /** The redeemer's own Ed25519 private key, hex (must match `redeemerEdPubHex`). */
  redeemerEdPrivHex: string
  /** The redeemer's own Ed25519 public key, hex — sent as `X-Starfish-Pub`. */
  redeemerEdPubHex: string
  /** HTTP method of the request being signed. */
  method: SignableMethod
  /** Path + query of the request being signed (e.g. `/pull/broadcast/abc`). */
  pathAndQuery: string
  /** Request body (signed). A JSON write signs its body; a blob upload signs empty. */
  body?: Uint8Array | string
  /** Host the request is bound to (folded into the signature). */
  host?: string
  /** Override signing timestamp (unix ms). Defaults to now. */
  ts?: number
  /** Override signing nonce (16 bytes). Defaults to fresh randomness. */
  nonce?: Uint8Array
}

/** Header set a client sends to redeem a public link. */
export interface RedeemHeaders {
  Authorization: string
  "X-Starfish-Sig": string
  "X-Starfish-Ts": string
  "X-Starfish-Nonce": string
  "X-Starfish-Pub": string
}

interface PublicLinkPayload {
  v: number
  cap: CapCert
}

// ── base64url (URL-fragment-safe) ──────────────────────────────────────────────
// Standard base64 (`+`, `/`, `=`) is unsafe in a URL fragment; map to base64url
// and strip padding. The exact same mapping is used by create + parse and is
// mirrored byte-for-byte in Python's `public_link.py`.

function b64urlEncode(s: string): string {
  const std = getBase64().encode(new TextEncoder().encode(s))
  return std.replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "")
}

function b64urlDecode(s: string): string {
  let std = s.replace(/-/g, "+").replace(/_/g, "/")
  const rem = std.length % 4
  if (rem === 2) std += "=="
  else if (rem === 3) std += "="
  else if (rem === 1) throw new Error("malformed public link: bad base64url length")
  return new TextDecoder().decode(getBase64().decode(std))
}

/** Encode a cap-cert for the `Authorization: Cap <…>` header (matches the resolver). */
function encodeCapAuth(cap: CapCert): string {
  return getBase64().encode(new TextEncoder().encode(stableStringify(cap as unknown as Record<string, unknown>)))
}

/**
 * Mint an audience cap and pack it into a shareable URL fragment.
 *
 * With `allowedIdentities` the link works only for those identities; without
 * it, any identity may redeem. Either way the link carries no private key.
 */
export async function createPublicLink(opts: CreatePublicLinkOpts): Promise<PublicLink> {
  const cap = await mintAudienceCap(opts.issEdPrivHex, opts.issEdPubHex, opts.collection, opts.scope, {
    audience: opts.allowedIdentities,
    expiresAt: opts.expiresAt,
    ttlSec: opts.ttlSec,
    nbf: opts.nbf,
    nonce: opts.nonce,
  })
  const payload: PublicLinkPayload = { v: PUBLIC_LINK_V, cap }
  const fragment = b64urlEncode(stableStringify(payload as unknown as Record<string, unknown>))
  return { fragment, cap }
}

/**
 * Decode and shape-check a public-link fragment. Does NOT verify the cap's
 * signature or expiry — the server does that at request time. Throws on a
 * malformed fragment, wrong payload version, or a non-audience / malformed cap.
 */
export function parsePublicLink(fragment: string): ParsedPublicLink {
  let payload: unknown
  try {
    payload = JSON.parse(b64urlDecode(fragment.trim()))
  } catch {
    throw new Error("malformed public link: invalid fragment encoding")
  }
  if (typeof payload !== "object" || payload === null) {
    throw new Error("malformed public link: payload is not an object")
  }
  const p = payload as { v?: unknown; cap?: unknown }
  if (p.v !== PUBLIC_LINK_V) {
    throw new Error(`unsupported public link version: ${String(p.v)}`)
  }
  const cap = p.cap as CapCert | undefined
  if (typeof cap !== "object" || cap === null || cap.kind !== "audience") {
    throw new Error("malformed public link: cap is not an audience cap")
  }
  // Structural validation only (no signature/expiry check — server's job).
  assertAudienceCapShape(cap)
  return { cap }
}

/**
 * Build the header set a redeemer sends with each request, signing with the
 * redeemer's own key and naming it via `X-Starfish-Pub`. Transport-agnostic:
 * the caller attaches these headers to its HTTP request.
 *
 * The body passed here MUST equal the bytes sent on the wire so the signature
 * the server reconstructs matches.
 */
export async function redeemPublicLink(
  parsed: ParsedPublicLink,
  opts: RedeemPublicLinkOpts,
): Promise<RedeemHeaders> {
  const req: SignableRequest = {
    method: opts.method,
    pathAndQuery: opts.pathAndQuery,
    body: opts.body,
    host: opts.host,
  }
  const { sig, ts, nonce } = await signRequest(req, opts.redeemerEdPrivHex, {
    ts: opts.ts,
    nonce: opts.nonce,
  })
  return {
    Authorization: `Cap ${encodeCapAuth(parsed.cap)}`,
    "X-Starfish-Sig": sig,
    "X-Starfish-Ts": String(ts),
    "X-Starfish-Nonce": nonce,
    "X-Starfish-Pub": opts.redeemerEdPubHex,
  }
}
